framework/src/userobject/ElementSubdomainModifier.C

```diff
@@ -0,0 +1,353 @@
+}
+
+void
+ElementSubdomainModifier::execute()
+{
+  // First, compute the desired subdomain ID for the current element.
+  SubdomainID subdomain_id = computeSubdomainID();
+
+  // If the current element's subdomain ID isn't what we want
+  if (subdomain_id != std::numeric_limits<SubdomainID>::max() &&
+      _current_elem->subdomain_id() != subdomain_id)
+  {
+    // Current element ID, used to index both the element on the displaced and undisplaced meshes.
+    dof_id_type elem_id = _current_elem->id();
+    // Change the element's subdomain
+    Elem * elem = _mesh.elemPtr(elem_id);
+    Elem * displaced_elem =
+        _displaced_problem ? _displaced_problem->mesh().elemPtr(elem_id) : nullptr;
+
+    // Save the affected nodes so that we can later update/initialize the solution
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+      _moved_nodes.insert(elem->node_id(i));
+
+    elem->subdomain_id() = subdomain_id;
+    _moved_elems.push_back(elem);
+    if (displaced_elem)
+    {
+      displaced_elem->subdomain_id() = subdomain_id;
+      _moved_displaced_elems.push_back(displaced_elem);
+    }
+  }
+}
+
+void
+ElementSubdomainModifier::finalize()
+{
+  /*
+    Synchronize ghost element subdomain ID
+    Note: this needs to be done before updating boundary info because
+    updating boundary requires the updated element subdomain ids
+  */
+  SyncSubdomainIds sync(_mesh.getMesh());
+  Parallel::sync_dofobject_data_by_id(_mesh.getMesh().comm(),
+                                      _mesh.getMesh().elements_begin(),
+                                      _mesh.getMesh().elements_end(),
+                                      sync);
+  updateBoundaryInfo(_mesh, _moved_elems);
+
+  // Similarly for the displaced mesh
+  if (_displaced_problem)
+  {
+    SyncSubdomainIds sync_displaced(_displaced_problem->mesh().getMesh());
+    Parallel::sync_dofobject_data_by_id(_displaced_problem->mesh().getMesh().comm(),
+                                        _displaced_problem->mesh().getMesh().elements_begin(),
+                                        _displaced_problem->mesh().getMesh().elements_end(),
+                                        sync_displaced);
+    updateBoundaryInfo(_displaced_problem->mesh(), _moved_displaced_elems);
+  }
+
+  // Reinit equation systems
+  _fe_problem.meshChanged();
+
+  // Apply initial condition for the newly moved elements and boundary nodes
+  buildMovedElemsRange();
+  buildMovedBndNodesRange();
+
+  if (_apply_ic)
+  {
+    _fe_problem.projectInitialConditionOnCustomRange(movedElemsRange(), movedBndNodesRange());
+
+    // Set old and older solution on the initialized dofs
+    setSolutionForMovedNodes(_fe_problem.getNonlinearSystemBase());
+    setSolutionForMovedNodes(_fe_problem.getAuxiliarySystem());
+    _fe_problem.restoreSolutions();
+  }
+
+  // Initialize stateful material properties for the newly activated elements
+  _fe_problem.initElementStatefulProps(movedElemsRange());
```

Comment:
I feel like it is better to  change "`moved`" to `activated`.  `activated` makes more sense for in this context

Reply:
I don't think so. The elements here really just changed subdomain IDs. `activated` would make perfect sense if we use this object for element activation, but would be misleading in the context of element deletion and moving interface.